File: ios/Sources/SettingsScreen.swift
import SwiftUI
import AVFoundation
import UserNotifications

struct SettingsScreen: View {
    let store: ServerStore
    @State private var isScanning = false
    @State private var manualURL = ""
    @State private var manualToken = ""
    @State private var manualName = ""
    @State private var renaming: Server?
    @State private var scanFailure: String?

    var body: some View {
        Form {
            Section {
                if store.servers.isEmpty {
                    Text("No servers yet. Scan the pairing code on your AppTesterClub instance, or enter the details below.")
                        .font(.system(size: 13, design: .rounded))
                        .foregroundStyle(Palette.textTertiary)
                } else {
                    ForEach(store.servers) { server in
                        Button {
                            store.select(server)
                        } label: {
                            HStack {
                                VStack(alignment: .leading, spacing: 2) {
                                    Text(server.displayName)
                                        .font(.system(size: 15, weight: .medium, design: .rounded))
                                        .foregroundStyle(Palette.textPrimary)
                                    Text(server.url.absoluteString)
                                        .font(.system(size: 11, design: .rounded))
                                        .foregroundStyle(Palette.textTertiary)
                                        .lineLimit(1)
                                        .truncationMode(.middle)
                                }
                                Spacer()
                                if store.selected?.id == server.id {
                                    Image(systemName: "checkmark")
                                        .foregroundStyle(Palette.accent)
                                }
                            }
                            .contentShape(.rect)
                        }
                        .swipeActions(edge: .leading) {
                            Button { renaming = server } label: {
                                Label("Rename", systemImage: "pencil")
                            }
                            .tint(Palette.accent)
                        }
                        .contextMenu {
                            Button { renaming = server } label: {
                                Label("Rename", systemImage: "pencil")
                            }
                        }
                    }
                    .onDelete { offsets in
                        offsets.map { store.servers[$0] }.forEach(store.remove)
                    }
                    .onMove { _, _ in }
                }
            } header: {
                Text("Servers")
            } footer: {
                Text("Several at once is supported: your own instance and a client's, each with its own builds.")
            }

            Section {
                Button {
                    scanFailure = nil
                    isScanning = true
                } label: {
                    Label("Scan pairing code", systemImage: "qrcode.viewfinder")
                }

                if let scanFailure {
                    Text(scanFailure)
                        .font(.system(size: 12, design: .rounded))
                        .foregroundStyle(Palette.danger)
                }
            } header: {
                Text("Add a server")
            } footer: {
                Text("Open /pair on your instance and scan what it shows. Beats typing a 48-character token on a phone.")
            }

            Section("Or enter it by hand") {
                TextField("Name (optional)", text: $manualName)
                    .autocorrectionDisabled()

                TextField("https://builds.example.com", text: $manualURL)
                    .textContentType(.URL)
                    .keyboardType(.URL)
                    .autocorrectionDisabled()
                    .textInputAutocapitalization(.never)

                SecureField("Access token", text: $manualToken)
                    .autocorrectionDisabled()
                    .textInputAutocapitalization(.never)

                Button("Add server") { addManual() }
                    .disabled(URL(string: manualURL)?.scheme != "https" || manualToken.isEmpty)
            }

            Section {
                LabeledContent("Permission") {
                    Text(permissionText)
                        .foregroundStyle(permissionColor)
                }
                LabeledContent("This device") {
                    Text(Push.shared.token == nil ? "Not registered" : "Registered")
                        .foregroundStyle(Push.shared.token == nil ? Palette.warning : Palette.success)
                }
                if Push.shared.token != nil {
                    Button("Re-register with all servers") {
                        Task { await Push.shared.register(with: store.servers) }
                    }
                }
            } header: {
                Text("Notifications")
            } footer: {
                Text("A build pushed to any server you are paired with arrives here. The server also needs an Apple push key configured, or nothing is sent.")
            }

            Section {
                LabeledContent("Version") {
                    Text(Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "—")
                        .monospacedDigit()
                }
                Link("Source and documentation",
                     destination: URL(string: "https://github.com/aditya-28/apptesterclub")!)
            } footer: {
                Text("AppTesterClub is open source. You built and signed this copy yourself, and it talks only to the servers listed above.")
            }
        }
        .navigationTitle("Settings")
        .navigationBarTitleDisplayMode(.inline)
        .task { await Push.shared.refreshAuthorization() }
        .sheet(item: $renaming) { server in
            RenameServerSheet(store: store, server: server)
        }
        .sheet(isPresented: $isScanning) {
            QRScanner { url in
                isScanning = false
                guard let server = PairingLink.parse(url) else {
                    scanFailure = "That is not an AppTesterClub pairing code, or its address is not HTTPS."
                    return
                }
                store.add(server)
            }
        }
    }

    private var permissionText: String {
        switch Push.shared.authorization {
        case .authorized, .provisional, .ephemeral: "Allowed"
        case .denied: "Denied in Settings"
        default: "Not asked yet"
        }
    }

    private var permissionColor: Color {
        switch Push.shared.authorization {
        case .authorized, .provisional, .ephemeral: Palette.success
        case .denied: Palette.danger
        default: Palette.textTertiary
        }
    }

    private func addManual() {
        guard let url = URL(string: manualURL), url.scheme == "https" else { return }
        let name = manualName.trimmingCharacters(in: .whitespacesAndNewlines)
        let server = Server(name: name.isEmpty ? (url.host() ?? "") : name,
                            url: url, token: manualToken)
        store.add(server)
        Task { await Push.shared.register(with: server) }
        manualURL = ""
        manualToken = ""
        manualName = ""
    }
}

/// Minimal QR reader. It wants exactly one payload — a pairing link — so it
/// hands back the first valid URL it sees and closes.
struct QRScanner: UIViewControllerRepresentable {
    let onFound: (URL) -> Void

    func makeUIViewController(context: Context) -> ScannerController {
        ScannerController(onFound: onFound)
    }

    func updateUIViewController(_ controller: ScannerController, context: Context) {}
}

@MainActor
final class ScannerController: UIViewController {
    private let onFound: (URL) -> Void
    private var delivered = false
    private var forwarder: MetadataForwarder?

    /// AVCaptureSession is documented as safe to start and stop from a
    /// background queue, which is required because both calls block. Swift 6
    /// cannot see that guarantee, hence the explicit opt-out rather than
    /// pretending the type is Sendable.
    private nonisolated(unsafe) let session = AVCaptureSession()
    private nonisolated(unsafe) static let queue = DispatchQueue(label: "club.apptester.capture")

    init(onFound: @escaping (URL) -> Void) {
        self.onFound = onFound
        super.init(nibName: nil, bundle: nil)
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) { fatalError("not used") }

    override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = .black

        guard
            let device = AVCaptureDevice.default(for: .video),
            let input = try? AVCaptureDeviceInput(device: device),
            session.canAddInput(input)
        else { return }
        session.addInput(input)

        let output = AVCaptureMetadataOutput()
        guard session.canAddOutput(output) else { return }
        session.addOutput(output)

        let forwarder = MetadataForwarder { [weak self] url in self?.deliver(url) }
        self.forwarder = forwarder
        output.setMetadataObjectsDelegate(forwarder, queue: .main)
        output.metadataObjectTypes = [.qr]

        let preview = AVCaptureVideoPreviewLayer(session: session)
        preview.frame = view.layer.bounds
        preview.videoGravity = .resizeAspectFill
        view.layer.addSublayer(preview)

        let session = self.session
        Self.queue.async { session.startRunning() }
    }

    override func viewDidDisappear(_ animated: Bool) {
        super.viewDidDisappear(animated)
        let session = self.session
        Self.queue.async { session.stopRunning() }
    }

    /// The capture output fires repeatedly while a code is in frame; the sheet
    /// should act on the first one only.
    private func deliver(_ url: URL) {
        guard !delivered else { return }
        delivered = true
        onFound(url)
    }
}

/// AVFoundation's delegate protocol is not main-actor isolated and a
/// UIViewController is, so the conformance cannot sit on the controller under
/// Swift 6. This forwarder is deliberately separate; the output queue is
/// `.main`, which `assumeIsolated` asserts rather than assumes.
private final class MetadataForwarder: NSObject, AVCaptureMetadataOutputObjectsDelegate {
    private let handler: @MainActor (URL) -> Void

    init(handler: @escaping @MainActor (URL) -> Void) {
        self.handler = handler
    }

    func metadataOutput(
        _ output: AVCaptureMetadataOutput,
        didOutput objects: [AVMetadataObject],
        from connection: AVCaptureConnection
    ) {
        guard
            let object = objects.first as? AVMetadataMachineReadableCodeObject,
            let value = object.stringValue,
            let url = URL(string: value)
        else { return }
        // Copy out of self first: capturing the property would send the whole
        // object across the isolation boundary, which Swift 6 rejects.
        let handler = self.handler
        MainActor.assumeIsolated { handler(url) }
    }
}

File: android/app/src/main/java/club/apptester/client/SettingsScreen.kt
package club.apptester.client

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardCapitalization
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.foundation.text.KeyboardOptions

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun SettingsScreen(model: BuildsViewModel, onBack: () -> Unit) {
    var name by remember { mutableStateOf("") }
    var url by remember { mutableStateOf("") }
    var token by remember { mutableStateOf("") }
    var renaming by remember { mutableStateOf<Server?>(null) }
    var revision by remember { mutableStateOf(0) }

    val servers = remember(revision) { model.servers.servers }

    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text("Settings", fontWeight = FontWeight.SemiBold) },
                navigationIcon = {
                    IconButton(onClick = onBack) {
                        Icon(Icons.Default.ArrowBack, contentDescription = "Back")
                    }
                },
            )
        },
    ) { padding ->
        LazyColumn(
            Modifier.padding(padding).fillMaxSize(),
            contentPadding = PaddingValues(16.dp),
            verticalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            item { Header("Servers") }

            if (servers.isEmpty()) {
                item {
                    Text(
                        "No servers yet. Scan the pairing code on your AppTesterClub instance, " +
                            "or enter the details below.",
                        fontSize = 13.sp,
                        color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.6f),
                    )
                }
            }

            items(servers, key = { it.id }) { server ->
                Row(
                    Modifier.fillMaxWidth()
                        .clickable { model.select(server); revision++ }
                        .padding(vertical = 10.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    Column(Modifier.weight(1f)) {
                        Text(server.displayName, fontWeight = FontWeight.Medium)
                        Text(server.subtitle, fontSize = 11.sp,
                            color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.6f))
                    }
                    if (model.servers.selected?.id == server.id) {
                        Icon(Icons.Default.Check, contentDescription = "Selected",
                            tint = Palette.accent)
                    }
                    IconButton(onClick = { renaming = server }) {
                        Icon(Icons.Default.Edit, contentDescription = "Rename")
                    }
                    IconButton(onClick = { model.servers.remove(server); revision++; model.refresh() }) {
                        Icon(Icons.Default.Delete, contentDescription = "Remove",
                            tint = Palette.danger)
                    }
                }
                HorizontalDivider()
            }

            item {
                Text(
                    "Several at once is supported: your own instance and a client's, " +
                        "each with its own builds.",
                    fontSize = 12.sp,
                    color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.6f),
                )
            }

            item { Header("Add a server") }
            item {
                OutlinedTextField(name, { name = it }, label = { Text("Name (optional)") },
                    modifier = Modifier.fillMaxWidth(), singleLine = true)
            }
            item {
                OutlinedTextField(url, { url = it },
                    label = { Text("https://builds.example.com") },
                    modifier = Modifier.fillMaxWidth(), singleLine = true,
                    keyboardOptions = KeyboardOptions(capitalization = KeyboardCapitalization.None))
            }
            item {
                OutlinedTextField(token, { token = it }, label = { Text("Access token") },
                    modifier = Modifier.fillMaxWidth(), singleLine = true,
                    visualTransformation = PasswordVisualTransformation(),
                    keyboardOptions = KeyboardOptions(capitalization = KeyboardCapitalization.None))
            }
            item {
                Button(
                    onClick = {
                        model.servers.add(Server(name = name.trim(), url = url.trim(), token = token.trim()))
                        name = ""; url = ""; token = ""; revision++
                        model.refresh()
                    },
                    enabled = url.trim().startsWith("https://") && token.isNotBlank(),
                    modifier = Modifier.fillMaxWidth(),
                ) { Text("Add server") }
            }
            item {
                Text(
                    "A pairing link from your instance's /pair page also works: open it on this " +
                        "phone and it fills all of this in.",
                    fontSize = 12.sp,
                    color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.6f),
                )
            }

            item { Header("About") }
            item {
                Text(
                    "AppTesterClub ${BuildConfig.VERSION_NAME} (${BuildConfig.VERSION_CODE})",
                    fontSize = 13.sp,
                )
                Text(
                    "Open source. You built and signed this copy yourself, and it talks only to " +
                        "the servers listed above.",
                    fontSize = 12.sp,
                    color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.6f),
                )
            }
        }
    }

    renaming?.let { server ->
        var text by remember(server.id) { mutableStateOf(server.name) }
        AlertDialog(
            onDismissRequest = { renaming = null },
            title = { Text("Name this server") },
            text = {
                OutlinedTextField(text, { text = it }, singleLine = true,
                    placeholder = { Text(server.subtitle) })
            },
            confirmButton = {
                TextButton(onClick = {
                    model.servers.rename(server, text); renaming = null; revision++
                }) { Text("Save") }
            },
            dismissButton = { TextButton(onClick = { renaming = null }) { Text("Cancel") } },
        )
    }
}

@Composable
private fun Header(text: String) {
    Text(
        text.uppercase(),
        fontSize = 11.sp,
        fontWeight = FontWeight.SemiBold,
        color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.5f),
        modifier = Modifier.padding(top = 14.dp),
    )
}
